From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RICE0707 <master.end.9377@gmail.com>
Date: Fri, 22 May 2026 09:43:47 +0800
Subject: [PATCH] purpur-port: Configurable minecart Place Anywhere settings


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 26060839dcaf9e92a1eb566b897d9053a46c978e..fea9ca04496c178a5234e05d609116cd7096d813 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -372,6 +372,8 @@ public class GlobalConfiguration extends ConfigurationPart {
         public boolean chestOpenWithBlockOnTop = false;
         @Comment("If true, death messages are only sent to the player who died instead of broadcasting to all players.")
         public boolean deathMessageOnlyBroadcastToAffectedPlayer = false;
+        @Comment("If true, minecarts can be placed on any block, not just rails.")
+        public boolean minecartPlaceAnywhere = false;
     }
     // purpur-port end
 
diff --git a/src/main/java/net/minecraft/world/item/MinecartItem.java b/src/main/java/net/minecraft/world/item/MinecartItem.java
index dd5a4463f4b0273f5c7491a3c78567e3ac6a3351..a5cf94246ca0cb9d902fcb6351f710345fd84966 100644
--- a/src/main/java/net/minecraft/world/item/MinecartItem.java
+++ b/src/main/java/net/minecraft/world/item/MinecartItem.java
@@ -120,8 +120,9 @@ public class MinecartItem extends Item {
         BlockState iblockdata = world.getBlockState(blockposition);
 
         if (!iblockdata.is(BlockTags.RAILS)) {
-            return InteractionResult.FAIL;
-        } else {
+            if (!io.papermc.paper.configuration.GlobalConfiguration.get().purpurPort.minecartPlaceAnywhere) return InteractionResult.FAIL; // purpur-port - place minecarts anywhere
+            if (iblockdata.isSolid()) blockposition = blockposition.relative(context.getClickedFace()); // purpur-port - place minecarts anywhere
+        } // purpur-port - place minecarts anywhere (removed else)
             ItemStack itemstack = context.getItemInHand();
 
             if (world instanceof ServerLevel) {
@@ -147,6 +148,6 @@ public class MinecartItem extends Item {
 
             itemstack.shrink(1);
             return InteractionResult.sidedSuccess(world.isClientSide);
-        }
+        // } // purpur-port - place minecarts anywhere
     }
-}
+}
\ No newline at end of file
